### ISSUE TITLE
Multi-core/Multithreaded support

### ISSUE THREAD

**Issue by FreeCAD-Bug-Importer** [FreeCAD/FreeCAD#5581]
Issue imported from https://tracker.freecad.org/view.php?id=1415

* **Reporter:** Pauvres_honteux
* **Date submitted:** 2/16/2014
* **FreeCAD version:** 
* **Category:** Feature
* **Status:** new
* **Tags:** multithreading

# Original report text

Hi, I'm not sure if FC currently is utilizing multiple cores/threads simultaneously, but it would be of huge benefit for the future when one could, at least in theory, move a feature (point, line, plane) in the resulting body and see the whole result updated in real time.

E.G. a casting whos shape depends upon an irregular hole pattern.
Or upon step changing a common draft angle parameter reflects its impact in real time.
Or moving holes and depressions in sheet metal and see the result in real time really enhances your understanding of that specific part you're working on.

And your customers will get a better understanding of how a particular change/demand will impact the end result which in the end cut costly time - a lot.

Regards Pauvres_honteux

# Additional information

Had this in the "Open discussion" before but it was apparently the wrong place.

# Other bug information

* **Priority:** normal
* **Severity:** minor
* **Category:** Feature
* **Updated:** 2/6/2021

# Discussion from Mantis ticket

### Comment by Pauvres_honteux 2014-02-23 18:22
This might be a starting point for learning parallel computing?
http://www.parallella.org/Introduction/

Regards Pauvres_honteux

---

### Comment by wmayer 2014-02-23 20:40
For parallel computing Qt already offers e.g. the (low-level) QThread class, QThreadPool or the high-level API QtConcurrent which we make use here and there.

The problem is not that we don't know how to parallelize it's more the lack of OCC to support it. At the moment there are only a few algorithms that are ready to use in separate threads.

---

### Comment by shoogen 2014-05-18 16:46
One improvemnt would be, that a recompute only blocks when applying changes to the GUI (changing QT elements or the scenegraph) and not while performing computations inside OCCT.

---

### Comment by Pauvres_honteux 2014-05-19 02:52
Hi Shoogen, I get a feeling there's a vital word (for me) missing in your comment [~4690] "...only blocks when applying..." ? Should there have been a word before or after "blocks" ?

---

### Comment by wbraun 2014-07-31 08:24
From my experience in an engineering company routinely working with complex models containing many parts, I strongly support parallel processing. The commercial CAD systems I know are all more or less single threaded Windows applications. As Processors these days no longer get faster, but have more cores instead, the only way to significantly improve productivity of CAD when working with complex projects seems to be to go parallel.
With Linux effortlessly scaling from an Arduino to supercomputers and an eight core workstation being available for 500 €, a truly parallel computing CAD system running on Linux could be a killer.
My experience with the commercial systems was that upgrading a several year old version on several year old PCs to the current version with carefully single-thred-performance-optimized PCs did not result in a noticeable performance increase. Loading and processing times of complex assemblies remained unchanged. A huge investment for nothing and very frustrating.

---

### Comment by Kunda1 2017-02-02 01:07
Linked https://tracker.freecad.org/view.php?id=2750

**Comment by luzpaz** [comment]
Found an interesting commit in OpenCascade git log:
ref: https://github.com/Open-Cascade-SAS/OCCT/commit/7e785937b3e65df86f35364e1036e446d6339cec

> 0025748: Parallel version of progress indicator
>
> Progress indication mechanism is refactored to support incrementing progress within multithreaded algorithms.
>
>The class Message_ProgressIndicator is only an interface to the user application.
>It accumulates the progress provided by progress scopes.
>The counter is protected by mutex for thread-safety.
>
>The new class Message_ProgressScope replacing Message_ProgressSentry should be used to advance the progress.
>The scopes are nested to each other to reflect the nested nature of operations.
>The new class Message_ProgressRange should be used to pass the progress to sub-scopes.
>
>All OCCT algorithms involving progress indication have been updated to new API.
>
>Improvements in Draw_ProgressIndicator:
> - Separate console mode has been added in order to make possible to put the progress into std::cout instead
  or in addition to the draw interpreter, instead of trigger option "-tclOutput".
> - Treatment of Ctrl-Break signal has been added.
  Now any operation can be aborted by Ctrl-C or Ctrl-Break keystroke.
>
> Added new test case 'perf fclasses progr_par' for testing of parallel work of the progress.

**Comment by luzpaz** [comment]
WIP for TechDraw hlr multithreading: https://forum.freecadweb.org/viewtopic.php?p=611590#p611590

**Comment by luzpaz** [comment]
Another upstream OCC multithreading commit: 
> JT Import - perform XT translation in multiple threads
> Creating mutex for proper parallel processing.

https://github.com/Open-Cascade-SAS/OCCT/commit/8da28014965697596aa3bde1c2827fe5c061468f

**Comment by luzpaz** [comment]
We need more devs looking at this. Working with upsteam OCC can be a pain-point though. Just take a look at the rest of the upstream OCC tickets.

**Comment by m1cha1** [comment]
Multithreaded STEP import is announced for the [upcoming 7.9 release.](https://dev.opencascade.org/content/open-cascade-development-insights-part-2)

**Comment by nazar-pc** [comment]
What kind of parallelism is this issue about? With over 300 constraints I already start feeling the bottlenecks of a single core processing, while the remaining 63 threads have very little work to do.

Probably not the place and/or time, but I'm working with Rust on a daily basis and parallelizing algorithms in Rust is much easier than in other languages while preserving safety and correctness.

**Comment by NXTler** [comment]
I think the biggest problems are Sketcher and Multi body transform. They both get very slow very quickly on only one core.

**Comment by xtemp09** [comment]
@nazar-pc, I have done only simple things in _FreeCAD_, no more complex than a vase or a headphone holder. A large project, involving 300 constraints, looks monstrous. Can you share an example file, so I run _FreeCAD_ with _gprof_ and, at least, take a look at the resultant flamegraph?

**Comment by nazar-pc** [comment]
It is just a sketch with some cutouts that need to be accurately positioned and that blows up quickly, especially when each cutout is a slot.

Here is the file: [ALotOfConstraints.zip](https://github.com/user-attachments/files/16477716/ALotOfConstraints.zip)
`MotherboardTray.FCStd` has a sketch `MotherboardTrayBase` with over 600 constraints in it.

When I add a new constraint it takes a couple of seconds to solve on slightly overclocked AMD 7970X (32C64T). If new constraint result in an error (like over-constrained) then it hangs for even longer.

I can see that solver is using exactly one thread right now.

**Comment by xtemp09** [comment]
nazar-pc, I've compiled _FreeCAD_ with `-g -O2` option and run under _callgrind_. _Eigen_ takes a lot of time, meaning that, the slowdown comes from an external library. I don't think that performance gain can be easily achieved.


If multithreading is possible, it must be implemented in _Eigen_.

<details><summary>Screenshot</summary>




![Sketcher](https://github.com/user-attachments/assets/3c4aace0-13a8-4eb6-a737-f83c20524628)



</details>

**Comment by nazar-pc** [comment]
Is there a chance that comment is old and things have changed since it was written?

**Comment by luzpaz** [comment]
9 year old commit (https://github.com/FreeCAD/FreeCAD/blame/270bb0122ea15bbc562942b9d0756fabb89362c0/src/Mod/Sketcher/App/planegcs/GCS.cpp#L497-L501)

**Comment by xtemp09** [comment]
I've compiled _FreeCAD_ with `-fopenmp -O2 -g`, modified the source to utilize 8 threads 👆, ran with 8 threads on 8-cored VM and performed the same operation in _Sketcher_. One can see that no increase in speed took place.




<details><summary>Before</summary>

![before](https://github.com/user-attachments/assets/b72cf155-3c1f-4b54-8ada-9ed28ebff93d)



</details>


<details><summary>After</summary>



![after](https://github.com/user-attachments/assets/27853155-356f-4c57-bbd0-f5a6329f370e)




</details>



```ruby
OS: Ubuntu 22.04.4 LTS (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.38387 (Git)
Build type: Release
Branch: main
Hash: c0c7bd647dc91f8807d8632585b6e6a53a040798
Python 3.10.12, Qt 5.15.14, Coin 4.0.2, Vtk 9.1.0, OCC 7.5.1
Locale: English/United States (en_US)
```

**Comment by nazar-pc** [comment]
Last release of Eigen was 3 years ago and given lack of multi-threading for crucial algorithms it might make sense to migrate to a different library that does take advantage of parallelism properly.

**Comment by dzid26** [comment]
It would be big improvement if 3D viewer manipulation was possible while things are computing.   Currently the whole GUI is blocked.

**Comment by tobiasfalk** [comment]
While reading through this I thought maybe the Assembly WB could benefit from multithreafing.
When recomputing a Assembly it also recomputes the separate part, and my thought is that the computing of those parts could be split in to there own threads, and only after all of them are finished the whole Assembly is computed. If this is not already done.

@PaddleStroke ?

**Comment by PaddleStroke** [comment]
Yes it sounds like multi-threading would be great (and possible) for assemblies. If the components do not depend on each others, I don't see why they could not be recomputed in parallel. Which would indeed reduce drastically the recompute time of large assemblies.

**Comment by tobiasfalk** [comment]
So, a work order would need to be generated based on the part dependencies. Similar how CMake/make looks what is dependent on what and compiels it based on that.

**Comment by reneas** [comment]
Since the forum posts linked [here](https://wiki.freecad.org/Multithreading) are all pretty old I'm coming here for the discussion.
From what I could read in those posts, there is still no real solution to the problem that some actions on linux cause single CPU cores to be at 100% changing in a random order.

As you can see here the cores 0, 4 , 6, 8 12 and 14 were all at 100% when i tried to just select an edge in the fillet tool (pretty complex model imported from a mesh and then converted to solid). Is there any way to enable multithread support for this kind of thing?

![Image](https://github.com/user-attachments/assets/437b74fa-a496-4654-9825-a75f8192d48e)

**Comment by oursland** [comment]
@reneas 

I'm not the authority in this part of code, but a lot of selection of items gets passed through the Python interpreter for the workbench, which at this time does somewhat enforce a single-threaded behavior.

A workaround for the pain of selecting the part you want is to disable preselection:

Navigate `Preferences`->`General`->`Selection` and uncheck `Enable preselection`

**Comment by reneas** [comment]
@oursland thanks for your reply!

I don't think preselection is the problem here because the CPU spikes only ocurr once I actually click on the edge or start an operation.

Preselecting when hovering over an entity works just fine.

**Comment by dzid26** [comment]
> some actions on linux cause single CPU cores to be at 100% **changing in a random order**

Why is that a problem? You can try to experiment with taskset and assigning [core affinity](https://www.google.com/search?q=freecad++core+affinity), but it's probably not going to make the difference. 

If you have particular issue with the fillet performance tool, you should create an issue (if it doesn't exist already) precisely describing the problem and attach the example.

**Comment by reneas** [comment]
I will definitely check out the core affinity settings, thanks for that!

My problem is that the processes are apparently not multi-threaded and therefor take a significant amount of time and I have the feeling that this could be sped up.

The problem doesn't only occur in the fillet but in a lot of things in the GUI and a lot of different operations. I'm pretty sure it's due to the fact that this model is quite big and complex (mesh converted to shape and then solid) but I feel like freecad should be able to handle this in a more modern way by utilizing more than just one of the 16 cores on my CPU.

Since I'm not super knowledgeable in how freecad handles things it might very well be that I'm suggesting something impossible. Please let me know then!

**Comment by luzpaz** [comment]
fillet has been a PITA for FreeCAD for years. Currently we're at the mercy of the OpenCasCade CAD kernel. Hence a new job has been posted to hire a liaison[^1] between the FreeCAD and OpenCasCade teams.

[^1]: https://blog.freecad.org/2025/01/24/job-openings-at-the-fpa/

**Comment by reneas** [comment]
As I said though, this doesn't only ocurr with fillets!

**Comment by luzpaz** [comment]
Indeed. There are some usual suspects (fillets, lofts etc... see https://github.com/FreeCAD/FreeCAD/labels/3rd%20party%3A%20OCC ). The bottleneck is OCC.

**Comment by oursland** [comment]
It should be noted that OpenCascade have moved development of [OCCT to GitHub](https://github.com/Open-Cascade-SAS/OCCT) and are now accepting PRs.  If a developer is interested in resolving an issue, they'll review and accept PRs.

**Comment by yorikvanhavre** [comment]
Don't you still need to sign some sort of developer agreement though?

**Comment by FlachyJoe** [comment]
@yorikvanhavre yes you need: [Open-Cascade-SAS/OCCT/CONTRIBUTING.md](https://github.com/Open-Cascade-SAS/OCCT/blob/cc30b9370013fd095f7d0cc8b8a159f6a1987ae2/.github/CONTRIBUTING.md)

**Comment by NXTler** [comment]
Is the completion of this issues relevant?
https://github.com/openscad/openscad/issues/391

**Comment by FlachyJoe** [comment]
@NXTler OpenSCAD is not OpenCascade

**Comment by luzpaz** [comment]
Part of  
* https://github.com/FreeCAD/FPA-grant-proposals/issues/36

**Comment by Doriangaensslen** [comment]
I would love to see some improvements there, because on my CAM jobs every click  on Operations "costs" about 15 minutes of waiting time. Of course I know that multithreaded coding is a whole other story, but still, when on 16 visible CPUs only one is calculating, it would mean a very big improvement. Or is this not related to Job Calculation in CAM?

**Comment by stefano2734** [comment]
A problem for decades.
Actual 7.9.3 and 8.0.1 are here available with announced thread safety by step import in 7.9.x.
So a new test is a good idea with 26.3 dev with libpack 3.4.0 or 3.5.5.
